src/DotNetty.Handlers/Timeout/IdleStateHandler.cs

```diff
@@ -0,0 +1,480 @@
+        readonly long writerIdleTimeNanos;
+        readonly long allIdleTimeNanos;
+
+        volatile IScheduledTask readerIdleTimeout;
+        long lastReadTime;
+        bool firstReaderIdleEvent = true;
+
+        volatile IScheduledTask writerIdleTimeout;
+        long lastWriteTime;
+        bool firstWriterIdleEvent = true;
+
+        volatile IScheduledTask allIdleTimeout;
+        bool firstAllIdleEvent = true;
+
+        volatile int state; // 0 - none, 1 - initialized, 2 - destroyed
+        volatile bool reading;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DotNetty.Handlers.Timeout.IdleStateHandler"/> class.
+        /// </summary>
+        /// <param name="readerIdleTimeSeconds">Reader idle time seconds.</param>
+        /// <param name="writerIdleTimeSeconds">Writer idle time seconds.</param>
+        /// <param name="allIdleTimeSeconds">All idle time seconds.</param>
+        public IdleStateHandler(
+            int readerIdleTimeSeconds,
+            int writerIdleTimeSeconds,
+            int allIdleTimeSeconds)
+            : this(TimeSpan.FromSeconds(readerIdleTimeSeconds), 
+                   TimeSpan.FromSeconds(writerIdleTimeSeconds), 
+                   TimeSpan.FromSeconds(allIdleTimeSeconds))
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DotNetty.Handlers.Timeout.IdleStateHandler"/> class.
+        /// </summary>
+        /// <param name="readerIdleTime">Reader idle time.</param>
+        /// <param name="writerIdleTime">Writer idle time.</param>
+        /// <param name="allIdleTime">All idle time.</param>
+        public IdleStateHandler(TimeSpan readerIdleTime, TimeSpan writerIdleTime, TimeSpan allIdleTime)
+        {
+            this.readerIdleTimeNanos = 
+                readerIdleTime.Ticks <= 0 ? 0 : Math.Max(TimeUtil.GetNanosFromTicks(readerIdleTime.Ticks), IdleStateHandler.MinTimeoutNanos);
+
+            this.writerIdleTimeNanos = 
+                writerIdleTime.Ticks <= 0 ? 0 : Math.Max(TimeUtil.GetNanosFromTicks(writerIdleTime.Ticks), IdleStateHandler.MinTimeoutNanos);
+
+            this.allIdleTimeNanos = 
+                allIdleTime.Ticks <= 0 ? 0 : Math.Max(TimeUtil.GetNanosFromTicks(allIdleTime.Ticks), IdleStateHandler.MinTimeoutNanos);
+
+            this.writeListener = new Action<Task>(antecedent =>
+                {
+                    Interlocked.Exchange(ref this.lastWriteTime, TimeUtil.GetSystemNanoTime());
+                    this.firstWriterIdleEvent = this.firstAllIdleEvent = true;
+                });
+        }
+
+        /// <summary>
+        /// Return the readerIdleTime that was given when instance this class in milliseconds.
+        /// </summary>
+        /// <returns>The reader idle time in millis.</returns>
+        public long GetReaderIdleTimeInMillis()
+        {
+            return TimeSpan.FromTicks(TimeUtil.GetTicksFromNanos(this.readerIdleTimeNanos)).Milliseconds;
```

Comment:
I don't see much reason to keep this quirky method. Let's replace it with a read-only property `TimeSpan ReaderIdleTime`.